lib/spack/spack/spec.py

```diff
@@ -2098,11 +2154,71 @@ def from_dict(data):
                 continue
 
             yaml_deps = node[name]['dependencies']
-            for dname, dhash, dtypes in Spec.read_yaml_dep_specs(yaml_deps):
+            for dname, dhash, dtypes, _ in Spec.read_yaml_dep_specs(yaml_deps):
                 deps[name]._add_dependency(deps[dname], dtypes)
 
         return spec
 
+    @staticmethod
+    def from_dict(data):
+        """Construct a spec from JSON/YAML.
+
+        Parameters:
+        data -- a nested dict/list data structure read from YAML or JSON.
+        """
+        # Check for the old format
+
+        if isinstance(data['spec'], list):
+            return Spec.from_old_dict(data)
+
+        # New Design
+
+        nodes = data['spec']['nodes']
+        hash_type = None
+        any_deps = False
+
+        # Pass 0: Determine hash type
+        for node in nodes:
+            if 'dependencies' in node.keys():
+                any_deps = True
+                for _, _, _, dhash_type in Spec.dependencies_from_node_dict(node):
+                    if dhash_type:
+                        hash_type = dhash_type
+                        break
+
+        if not any_deps:  # If we never see a dependency...
+            hash_type = '_hash'  # guess
```

Comment:
I don't think this is a guess -- it really doesn't matter which it is at this point, right?

Reply:
I also updated this to use the `attr` for consistency.